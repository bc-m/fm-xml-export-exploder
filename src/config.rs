#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum CatalogType {
    Accounts,
    BaseDirectory,
    BaseTable,
    CalcsForCustomFunctions,
    CustomFunctions,
    CustomMenu,
    CustomMenuSet,
    ExternalDataSource,
    ExtendedPrivileges,
    FieldsForTables,
    FileAccess,
    Layout,
    Library,
    OptionsForValueLists,
    PrivilegeSets,
    Relationship,
    Script,
    StepsForScripts,
    TableOccurrence,
    Theme,
    ValueList,
}

impl CatalogType {
    pub fn from_bytes(bytes: &[u8]) -> Option<Self> {
        match bytes {
            b"AccountsCatalog" => Some(Self::Accounts),
            b"BaseDirectoryCatalog" => Some(Self::BaseDirectory),
            b"BaseTableCatalog" => Some(Self::BaseTable),
            b"CalcsForCustomFunctions" => Some(Self::CalcsForCustomFunctions),
            b"CustomFunctionsCatalog" => Some(Self::CustomFunctions),
            b"CustomMenuCatalog" => Some(Self::CustomMenu),
            b"CustomMenuSetCatalog" => Some(Self::CustomMenuSet),
            b"ExternalDataSourceCatalog" => Some(Self::ExternalDataSource),
            b"ExtendedPrivilegesCatalog" => Some(Self::ExtendedPrivileges),
            b"FieldsForTables" => Some(Self::FieldsForTables),
            b"FileAccessCatalog" => Some(Self::FileAccess),
            b"LayoutCatalog" => Some(Self::Layout),
            b"LibraryCatalog" => Some(Self::Library),
            b"OptionsForValueLists" => Some(Self::OptionsForValueLists),
            b"PrivilegeSetsCatalog" => Some(Self::PrivilegeSets),
            b"RelationshipCatalog" => Some(Self::Relationship),
            b"ScriptCatalog" => Some(Self::Script),
            b"StepsForScripts" => Some(Self::StepsForScripts),
            b"TableOccurrenceCatalog" => Some(Self::TableOccurrence),
            b"ThemeCatalog" => Some(Self::Theme),
            b"ValueListCatalog" => Some(Self::ValueList),
            _ => None,
        }
    }

    pub fn get_config(&self) -> CatalogConfig {
        match self {
            Self::Accounts => CatalogConfig {
                catalog_item_name: b"Account".to_vec(),
                out_folder_name: "accounts".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "Account/AccountReference/@id".to_string(),
            },
            Self::BaseDirectory => CatalogConfig {
                catalog_item_name: b"BaseDirectory".to_vec(),
                out_folder_name: "base_directory".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::BaseTable => CatalogConfig {
                catalog_item_name: b"BaseTable".to_vec(),
                out_folder_name: "table_stubs".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::CalcsForCustomFunctions => CatalogConfig {
                catalog_item_name: b"CustomFunctionCalc".to_vec(),
                out_folder_name: "custom_function_calcs".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "CustomFunctionCalc/CustomFunctionReference/@id".to_string(),
            },
            Self::CustomFunctions => CatalogConfig {
                catalog_item_name: b"CustomFunction".to_vec(),
                out_folder_name: "custom_function_stubs".to_string(),
                wrapped_in_object_list: true,
                uses_folders: true,
                id_path: "".to_string(),
            },
            Self::CustomMenu => CatalogConfig {
                catalog_item_name: b"CustomMenu".to_vec(),
                out_folder_name: "custom_menus".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::CustomMenuSet => CatalogConfig {
                catalog_item_name: b"CustomMenuSet".to_vec(),
                out_folder_name: "custom_menu_sets".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::ExternalDataSource => CatalogConfig {
                catalog_item_name: b"ExternalDataSource".to_vec(),
                out_folder_name: "external_data_sources".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::ExtendedPrivileges => CatalogConfig {
                catalog_item_name: b"ExtendedPrivilege".to_vec(),
                out_folder_name: "extended_privileges".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::FieldsForTables => CatalogConfig {
                catalog_item_name: b"FieldCatalog".to_vec(),
                out_folder_name: "table_fields".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "FieldCatalog/BaseTableReference/@id".to_string(),
            },
            Self::FileAccess => CatalogConfig {
                catalog_item_name: b"Authorization".to_vec(),
                out_folder_name: "file_access".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::Layout => CatalogConfig {
                catalog_item_name: b"Layout".to_vec(),
                out_folder_name: "layouts".to_string(),
                wrapped_in_object_list: false,
                uses_folders: true,
                id_path: "Layout/LayoutReference/@id".to_string(),
            },
            Self::Library => CatalogConfig {
                catalog_item_name: b"BinaryData".to_vec(),
                out_folder_name: "libraries".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "BinaryData/LibraryReference/@id".to_string(),
            },
            Self::OptionsForValueLists => CatalogConfig {
                catalog_item_name: b"ValueList".to_vec(),
                out_folder_name: "value_list_options".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "ValueList/ValueListReference/@id".to_string(),
            },
            Self::PrivilegeSets => CatalogConfig {
                catalog_item_name: b"PrivilegeSet".to_vec(),
                out_folder_name: "privilege_sets".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::Relationship => CatalogConfig {
                catalog_item_name: b"Relationship".to_vec(),
                out_folder_name: "relationships".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::Script => CatalogConfig {
                catalog_item_name: b"Script".to_vec(),
                out_folder_name: "script_stubs".to_string(),
                wrapped_in_object_list: false,
                uses_folders: true,
                id_path: "".to_string(),
            },
            Self::StepsForScripts => CatalogConfig {
                catalog_item_name: b"Script".to_vec(),
                out_folder_name: "script_steps".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "Script/ScriptReference/@id".to_string(),
            },
            Self::TableOccurrence => CatalogConfig {
                catalog_item_name: b"TableOccurrence".to_vec(),
                out_folder_name: "table_occurrences".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::Theme => CatalogConfig {
                catalog_item_name: b"Theme".to_vec(),
                out_folder_name: "themes".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::ValueList => CatalogConfig {
                catalog_item_name: b"ValueList".to_vec(),
                out_folder_name: "value_list_stubs".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
        }
    }
}

#[derive(Debug)]
pub struct CatalogConfig {
    pub catalog_item_name: Vec<u8>,
    pub out_folder_name: String,
    pub wrapped_in_object_list: bool,
    pub uses_folders: bool,
    pub id_path: String,
}

pub struct Flags {
    pub parse_all_lines: bool,
    pub lossless: bool,
    pub verbose: Option<Option<String>>,
}
