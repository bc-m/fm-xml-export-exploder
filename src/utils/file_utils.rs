use anyhow::{anyhow, Error, Result};
use std::path::PathBuf;

/// Validate that a path exists and is a directory
pub fn valid_dir_or_throw(dir_path: &PathBuf) -> Result<(), Error> {
    let metadata = std::fs::metadata(dir_path)
        .map_err(|_| anyhow!("Path '{}' not exists", dir_path.display()))?;

    match metadata.is_dir() {
        true => Ok(()),
        false => Err(anyhow!("Path '{}' is not a directory", dir_path.display())),
    }
}

/// Join scope ID and name into a formatted string
pub fn join_scope_id_and_name(scope_id: &str, scope_name: &str) -> String {
    format!("{} - ID {}", scope_name, scope_id)
}

/// Escape filename by replacing special characters with underscores
pub fn escape_filename(filename: &str) -> String {
    let escaped_filename = filename
        .chars()
        .map(|c| match c {
            '/' | '\\' | ':' | '*' | '?' | '"' | '<' | '>' | '|' | '{' => '_',
            _ => c,
        })
        .collect::<String>();

    escaped_filename.trim().to_string()
}

/// Check if a line should be skipped based on content
pub fn should_skip_line(line: &str) -> bool {
    if line.contains("<TagList></TagList>") {
        return true;
    };
    if line.contains("<OwnerID></OwnerID>") {
        return true;
    };
    if line.contains("<Options>0</Options>") {
        return true;
    }; // Default
    if line.contains("<Options>1048576</Options>") {
        return true;
    }; // Breakpoint
    if line.contains("<UUID>") && line.contains("</UUID>") {
        return true;
    };

    false
}
