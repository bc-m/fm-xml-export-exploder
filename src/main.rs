use std::{fs, path::PathBuf, time::Instant};

use anyhow::Result;
use clap::Parser;
use rayon::prelude::*;

use crate::config::Flags;
use crate::utils::file_utils::valid_dir_or_throw;
use crate::utils::xml_utils::XmlEventType;
use crate::xml_processor::explode_xml;

mod catalog;
mod config;
mod script_sanitizer;
mod script_steps;
mod supporting;
mod tests;
mod utils;
mod xml_processor;

/// Parse all as XML exported FileMaker solutions from source directory and explode them to target directory.
#[derive(Parser)]
#[command(version, about, long_about = None)]
struct Args {
    /// The source directory to read input
    source: PathBuf,

    /// The target directory to write output
    target: PathBuf,

    /// Parse all lines (or skip less important ones to reduce noise)
    #[arg(short, long)]
    all_lines: bool,

    /// Retain all information from the main xml (or skip less important catalogs and attributes)
    #[arg(short, long)]
    lossless: bool,

    /// Enable debug/verbose output. Optionally filter by catalog name.
    /// Example: --verbose "CustomFunctionsCatalog"
    #[arg(short, long, value_name = "FILTER", num_args(0..=1))]
    verbose: Option<Option<String>>,
}

#[derive(Debug, Default)]
pub struct Skeleton {
    pub content: String,
    pub previous_line: String,
    pub previous_event_type: XmlEventType,
}

fn main() -> Result<()> {
    let start = Instant::now();

    let args = Args::parse();
    let in_dir = args.source;
    let out_dir = args.target;
    let flags = Flags {
        parse_all_lines: args.all_lines,
        lossless: args.lossless,
        verbose: args.verbose,
    };

    match &flags.verbose {
        Some(Some(filter)) => println!("🔍 Verbose mode (filtered on {})", filter),
        Some(None) => println!("🔊 Verbose mode (all)"),
        None => (), // silent
    };

    valid_dir_or_throw(&in_dir)?;

    // Read directory contents
    let paths = fs::read_dir(in_dir)?
        .filter_map(|entry| entry.ok().map(|e| e.path())) // Filter out directories and unwrap results
        .filter(|path| path.is_file() && path.extension().unwrap_or_default() == "xml") // Filter XML files
        .collect::<Vec<_>>(); // Collect paths into a vector

    println!("Start processing {} files...", paths.len());

    // Process XML files in parallel
    paths.par_iter().for_each(|path| {
        match explode_xml(path, &out_dir, &flags) {
            Ok(_) => {}
            Err(err) => {
                let file_name = path.file_name().unwrap().to_str().unwrap();
                eprintln!("Failed to process file '{}': {}", file_name, err)
            }
        };
    });

    let duration = start.elapsed();
    if duration.as_secs() > 9 {
        println!("Completed in {:?} seconds.", duration.as_secs());
    } else {
        println!("Completed in {:?} ms.", duration.as_millis());
    }

    Ok(())
}
