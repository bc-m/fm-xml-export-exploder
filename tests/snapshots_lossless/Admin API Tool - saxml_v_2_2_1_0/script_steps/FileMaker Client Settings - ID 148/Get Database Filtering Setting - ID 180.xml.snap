---
source: src/main.rs
expression: output_content
---
<Script>
	<ScriptReference id="180" name="Get Database Filtering Setting" UUID="57598923-F1E2-49A2-948E-96E6E244DE38"></ScriptReference>
	<ObjectList membercount="36">
		<Step index="0" id="89" name="# (Kommentar)" enable="True">
			<UUID>B851C9A7-C44B-4772-A9EF-082D60327ADA</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value=" PURPOSE: Filter List of Databases"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="1" id="89" name="# (Kommentar)" enable="True">
			<UUID>3E9DB09F-49F8-4B96-9A16-3DDBB2C53C02</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value=" HISTORY:&#13;   - Created:   2018-05-09 by Mike Duncan (mduncan@soliantconsulting.com)&#13;   - Modified:  2019-04-23 by Mike Duncan, update to v2 admin api"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="2" id="89" name="# (Kommentar)" enable="True">
			<UUID>FC864B12-28C1-48C8-A544-9D27CCE449FB</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="3" id="85" name="AnwenderAbbruchZulassen setzen" enable="True">
			<UUID>6D9E3B3F-4645-43B4-9161-DC5A0CB027A5</UUID>
			<OwnerID></OwnerID>
			<Options>196608</Options>
			<ParameterValues membercount="1">
				<Parameter type="Boolean">
					<Boolean id="131072" value="True"></Boolean>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="4" id="86" name="Fehleraufzeichnung setzen" enable="True">
			<UUID>9A276859-2EC7-4D0C-8BF4-5DD551D9B17C</UUID>
			<OwnerID></OwnerID>
			<Options>196608</Options>
			<ParameterValues membercount="1">
				<Parameter type="Boolean">
					<Boolean id="131072" value="True"></Boolean>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="5" id="141" name="Variable setzen" enable="True">
			<UUID>BD0C689A-D807-465B-9E20-290A84F3F21D</UUID>
			<OwnerID></OwnerID>
			<Options>16388</Options>
			<ParameterValues membercount="1">
				<Parameter type="Variable">
					<value>
						<Calculation datatype="1" position="1">
							<Calculation>
								<Text><![CDATA[0]]></Text>
								<ChunkList hash="9BF55A5D51230A8EA20D4BAA0F468EEC">
									<Chunk type="NoRef">0</Chunk>
								</ChunkList>
							</Calculation>
						</Calculation>
					</value>
					<Name value="$error"></Name>
					<repetition></repetition>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="6" id="89" name="# (Kommentar)" enable="True">
			<UUID>62078F79-D799-48C6-9F06-BC2D59A294A6</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="7" id="89" name="# (Kommentar)" enable="True">
			<UUID>79C8C7F9-E5B6-4633-8CBF-DA676181A7B5</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value="-------------------------------------------------------------"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="8" id="89" name="# (Kommentar)" enable="True">
			<UUID>1CC9E31A-9B97-4516-8417-F708E16BA513</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value=" SET-UP"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="9" id="89" name="# (Kommentar)" enable="True">
			<UUID>2F4672BD-B5DE-4E35-8492-A313371DA56C</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value="-------------------------------------------------------------"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="10" id="1" name="Script ausführen" enable="True">
			<UUID>1B85501E-FD3E-4AD1-968B-A5DE70CFB34C</UUID>
			<OwnerID></OwnerID>
			<Options>64</Options>
			<ParameterValues membercount="2">
				<Parameter type="List">
					<List name="Aus Liste" value="1">
						<ScriptReference id="35" name="Set Config" UUID="6E806D84-9200-44AE-B032-89A6016326FF"></ScriptReference>
					</List>
				</Parameter>
				<Parameter type="Parameter">
					<Parameter></Parameter>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="11" id="141" name="Variable setzen" enable="True">
			<UUID>FFCF4BA2-0C69-4B65-B9B8-A8954ED10071</UUID>
			<OwnerID></OwnerID>
			<Options>16388</Options>
			<ParameterValues membercount="1">
				<Parameter type="Variable">
					<value>
						<Calculation datatype="1" position="1">
							<Calculation>
								<Text><![CDATA["/fmclients/databasevisibility"]]></Text>
								<ChunkList hash="460798E4B16F71A5CBE1FA6492A49A22">
									<Chunk type="NoRef">&quot;/fmclients/databasevisibility&quot;</Chunk>
								</ChunkList>
							</Calculation>
						</Calculation>
					</value>
					<Name value="$endpoint"></Name>
					<repetition></repetition>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="12" id="141" name="Variable setzen" enable="True">
			<UUID>502F7F8E-2140-4438-B4F2-190BFA5D92DA</UUID>
			<OwnerID></OwnerID>
			<Options>16388</Options>
			<ParameterValues membercount="1">
				<Parameter type="Variable">
					<value>
						<Calculation datatype="1" position="1">
							<Calculation>
								<Text><![CDATA[JSONSetElement ( ""   ;[ "databaseVisibility"; FMS__fmserver::databaseVisibility ; JSONBoolean ] )]]></Text>
								<ChunkList hash="FA309FE03ACF7D2600BF6F8C72D61657">
									<Chunk type="FunctionRef">JSONSetElement</Chunk>
									<Chunk type="NoRef"> ( &quot;&quot; &#13;  ;[ &quot;databaseVisibility&quot;; </Chunk>
									<Chunk type="FieldRef">
										<FieldReference id="152" name="databaseVisibility" repetition="1" UUID="FCB33E3D-4670-4580-B937-F0ACD4E3FE30">
											<TableOccurrenceReference id="1065089" name="FMS__fmserver" UUID="597569C5-24CD-486D-84D6-D79B9D3BDD5B"></TableOccurrenceReference>
										</FieldReference>
									</Chunk>
									<Chunk type="NoRef"> ; </Chunk>
									<Chunk type="FunctionRef">JSONBoolean</Chunk>
									<Chunk type="NoRef"> ]&#13; )</Chunk>
								</ChunkList>
							</Calculation>
						</Calculation>
					</value>
					<Name value="$json"></Name>
					<repetition></repetition>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="13" id="141" name="Variable setzen" enable="True">
			<UUID>D6D62408-FE45-4607-BBB9-10AF0BA2257F</UUID>
			<OwnerID></OwnerID>
			<Options>16388</Options>
			<ParameterValues membercount="1">
				<Parameter type="Variable">
					<value>
						<Calculation datatype="1" position="1">
							<Calculation>
								<Text><![CDATA[" --request GET" &" --header \"Authorization: Bearer " & $$my.token & "\"" &" --header \"Content-Type: application/json\"" &" --dump-header $$my.headers"]]></Text>
								<ChunkList hash="79E981652F1EB4C1A038BEF0CCFB8B63">
									<Chunk type="NoRef">&quot; --request GET&quot; &amp;&#13;&quot; --header \&quot;Authorization: Bearer &quot; &amp; </Chunk>
									<Chunk type="VariableReference">$$my.token</Chunk>
									<Chunk type="NoRef"> &amp; &quot;\&quot;&quot; &amp;&#13;&quot; --header \&quot;Content-Type: application/json\&quot;&quot; &amp;&#13;&#13;&quot; --dump-header $$my.headers&quot;</Chunk>
								</ChunkList>
							</Calculation>
						</Calculation>
					</value>
					<Name value="$curloptions"></Name>
					<repetition></repetition>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="14" id="89" name="# (Kommentar)" enable="True">
			<UUID>3F498512-8F9A-4B40-8FD7-F4EE0D48BD07</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="15" id="89" name="# (Kommentar)" enable="True">
			<UUID>FC89DEB9-0046-44F5-965B-B4E2E3C1CFF7</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value="-------------------------------------------------------------"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="16" id="89" name="# (Kommentar)" enable="True">
			<UUID>4DF850AC-BB17-480E-8920-C33A893DA819</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value=" PROCESS"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="17" id="89" name="# (Kommentar)" enable="True">
			<UUID>AA49CE54-5D7C-4390-83F8-E9234E2EE1C2</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value="-------------------------------------------------------------"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="18" id="160" name="Aus URL einfügen" enable="True">
			<UUID>806AFA4F-E3F9-461F-9DA2-AFFB63A112E5</UUID>
			<OwnerID></OwnerID>
			<Options>20613</Options>
			<ParameterValues membercount="6">
				<Parameter type="Boolean">
					<Boolean type="SSL-Zertifikate verifizieren" id="268435456" value="False"></Boolean>
				</Parameter>
				<Parameter type="Boolean">
					<Boolean type="Auswahl" id="4096" value="True"></Boolean>
				</Parameter>
				<Parameter type="Boolean">
					<Boolean type="Mit Dialog" id="128" value="False"></Boolean>
				</Parameter>
				<Parameter type="Target">
					<Variable value="$$my.result">
						<repetition value="1"></repetition>
					</Variable>
				</Parameter>
				<Parameter type="URL">
					<URL autoEncode="True">
						<Calculation datatype="1" position="0">
							<Calculation>
								<Text><![CDATA[$$server & $endpoint]]></Text>
								<ChunkList hash="9EA92AEF077FA44AE68308694556F8CB">
									<Chunk type="VariableReference">$$server</Chunk>
									<Chunk type="NoRef"> &amp; </Chunk>
									<Chunk type="VariableReference">$endpoint</Chunk>
								</ChunkList>
							</Calculation>
						</Calculation>
					</URL>
				</Parameter>
				<Parameter type="Calculation">
					<Calculation datatype="1" position="1">
						<Calculation>
							<Text><![CDATA[$curloptions]]></Text>
							<ChunkList hash="205B4DA1BAD160DEDA31FE70167BBA78">
								<Chunk type="VariableReference">$curloptions</Chunk>
							</ChunkList>
						</Calculation>
					</Calculation>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="19" id="1" name="Script ausführen" enable="True">
			<UUID>F21323BD-24D2-4DAC-9A74-800C95108F90</UUID>
			<OwnerID></OwnerID>
			<Options>64</Options>
			<ParameterValues membercount="2">
				<Parameter type="List">
					<List name="Aus Liste" value="1">
						<ScriptReference id="31" name="Handle response" UUID="44FA18ED-6DEA-4F32-BDDB-09F438F6D005"></ScriptReference>
					</List>
				</Parameter>
				<Parameter type="Parameter">
					<Parameter></Parameter>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="20" id="89" name="# (Kommentar)" enable="True">
			<UUID>E325298C-21C5-4D94-8C57-A9095A9DD1C2</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="21" id="68" name="Wenn" enable="True">
			<UUID>40A81FDB-7485-4FA2-8D5C-BC4B9E087377</UUID>
			<OwnerID></OwnerID>
			<Options>16384</Options>
			<ParameterValues membercount="1">
				<Parameter type="Calculation">
					<Calculation datatype="7" position="0">
						<Calculation>
							<Text><![CDATA[Hole ( ScriptErgebnis ) = "1"]]></Text>
							<ChunkList hash="8C6CBCB21CCFFBA5E461DF278EC4FA0D">
								<Chunk type="FunctionRef">Hole</Chunk>
								<Chunk type="NoRef"> ( </Chunk>
								<Chunk type="FunctionRef">ScriptErgebnis</Chunk>
								<Chunk type="NoRef"> ) = &quot;1&quot;</Chunk>
							</ChunkList>
						</Calculation>
					</Calculation>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="22" id="89" name="# (Kommentar)" enable="True">
			<UUID>F20958AB-E798-4B9B-AFAD-15FB94BAE74D</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value="parse json response"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="23" id="76" name="Feldwert setzen" enable="True">
			<UUID>630B5CBD-5ED0-4528-835C-988ED97A0AED</UUID>
			<OwnerID></OwnerID>
			<Options>16385</Options>
			<ParameterValues membercount="2">
				<Parameter type="FieldReference">
					<FieldReference id="152" name="databaseVisibility" UUID="FCB33E3D-4670-4580-B937-F0ACD4E3FE30">
						<repetition value="1"></repetition>
						<TableOccurrenceReference id="1065089" name="FMS__fmserver" UUID="597569C5-24CD-486D-84D6-D79B9D3BDD5B"></TableOccurrenceReference>
					</FieldReference>
				</Parameter>
				<Parameter type="Calculation">
					<Calculation datatype="1" position="0">
						<Calculation>
							<Text><![CDATA[JSONGetElement ( $$my.result ; "response.databaseVisibility" )]]></Text>
							<ChunkList hash="A17F7EB50E5F1D1E1673AC903F9D5397">
								<Chunk type="FunctionRef">JSONGetElement</Chunk>
								<Chunk type="NoRef"> ( </Chunk>
								<Chunk type="VariableReference">$$my.result</Chunk>
								<Chunk type="NoRef"> ; &quot;response.databaseVisibility&quot; )</Chunk>
							</ChunkList>
						</Calculation>
					</Calculation>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="24" id="89" name="# (Kommentar)" enable="True">
			<UUID>42F2478D-0B70-4FB0-AD07-337C059386ED</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="25" id="70" name="Ende (wenn)" enable="True">
			<UUID>FDF69BFD-FEFA-4D62-A11C-2BD9A369857D</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
		</Step>
		<Step index="26" id="89" name="# (Kommentar)" enable="True">
			<UUID>686F2A4A-8A51-42D0-AE1E-C61769B6CC55</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value="-------------------------------------------------------------"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="27" id="89" name="# (Kommentar)" enable="True">
			<UUID>EE156142-0C3B-404C-86E1-86AC3AA80AD5</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value=" CLEAN UP"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="28" id="89" name="# (Kommentar)" enable="True">
			<UUID>E9DA2BD8-63CB-4F89-AD86-75CE658DAE3A</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value="-------------------------------------------------------------"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="29" id="68" name="Wenn" enable="True">
			<UUID>E46EF983-A62E-487B-B79E-C71C798A308E</UUID>
			<OwnerID></OwnerID>
			<Options>16384</Options>
			<ParameterValues membercount="1">
				<Parameter type="Calculation">
					<Calculation datatype="7" position="0">
						<Calculation>
							<Text><![CDATA[$error = 0]]></Text>
							<ChunkList hash="CE3C173E2CD2554CBB236CEE7C56CBB6">
								<Chunk type="VariableReference">$error</Chunk>
								<Chunk type="NoRef"> = 0</Chunk>
							</ChunkList>
						</Calculation>
					</Calculation>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="30" id="89" name="# (Kommentar)" enable="True">
			<UUID>1D9AF709-FE4D-4AF8-A070-00FA87DB61D6</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value=" set to exit true if you do not need to pass an error"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="31" id="141" name="Variable setzen" enable="True">
			<UUID>297A768B-9958-4219-8C58-D95DAE3AFC57</UUID>
			<OwnerID></OwnerID>
			<Options>16388</Options>
			<ParameterValues membercount="1">
				<Parameter type="Variable">
					<value>
						<Calculation datatype="1" position="1">
							<Calculation>
								<Text><![CDATA[1 // exit true]]></Text>
								<ChunkList hash="DC134DF247804ECF23193E2024A59A1B">
									<Chunk type="NoRef">1 // exit true</Chunk>
								</ChunkList>
							</Calculation>
						</Calculation>
					</value>
					<Name value="$error"></Name>
					<repetition></repetition>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="32" id="70" name="Ende (wenn)" enable="True">
			<UUID>AAF4244E-CD20-4E26-A194-37A8DF34FC00</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
		</Step>
		<Step index="33" id="89" name="# (Kommentar)" enable="True">
			<UUID>50A472AC-C546-48B4-AFDA-A639D9973CC4</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="34" id="103" name="Aktuelles Script verlassen" enable="True">
			<UUID>A95B1E69-54BA-4305-B005-3ADD6C062C22</UUID>
			<OwnerID></OwnerID>
			<Options>16384</Options>
			<ParameterValues membercount="1">
				<Parameter type="Calculation">
					<Calculation datatype="1" position="0">
						<Calculation>
							<Text><![CDATA[$error]]></Text>
							<ChunkList hash="926F59D0E89012933FB1C2137ABF78F1">
								<Chunk type="VariableReference">$error</Chunk>
							</ChunkList>
						</Calculation>
					</Calculation>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="35" id="89" name="# (Kommentar)" enable="True">
			<UUID>0D1519CA-C74A-492A-9080-824A243B4F14</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
	</ObjectList>
</Script>
